phase2/src/parameters.rs

```diff
@@ -658,188 +286,232 @@ impl MPCParameters {
     }
+    Ok(())
+}
 
-    // A/B_G1/B_G2 doesn't change at all
-    if before.params.a != after.params.a {
-        return Err(());
-    }
-    if before.params.b_g1 != after.params.b_g1 {
-        return Err(());
+fn ensure_unchanged_vec<T: PartialEq>(
+    before: &[T],
+    after: &[T],
+    kind: InvariantKind,
+) -> Result<()> {
+    if before.len() != after.len() {
+        return Err(Phase2Error::InvalidLength.into());
     }
-    if before.params.b_g2 != after.params.b_g2 {
-        return Err(());
+    for (before, after) in before.iter().zip(after) {
+        ensure_unchanged(before, after, kind.clone())?
     }
+    Ok(())
+}
 
-    // alpha/beta/gamma don't change
-    if before.params.vk.alpha_g1 != after.params.vk.alpha_g1 {
-        return Err(());
-    }
-    if before.params.vk.beta_g1 != after.params.vk.beta_g1 {
-        return Err(());
-    }
-    if before.params.vk.beta_g2 != after.params.vk.beta_g2 {
-        return Err(());
-    }
-    if before.params.vk.gamma_g2 != after.params.vk.gamma_g2 {
-        return Err(());
+fn ensure_unchanged<T: PartialEq>(before: T, after: T, kind: InvariantKind) -> Result<()> {
+    if before != after {
+        return Err(Phase2Error::BrokenInvariant(kind).into());
     }
+    Ok(())
+}
 
-    // IC shouldn't change, as gamma doesn't change
-    if before.params.vk.ic != after.params.vk.ic {
-        return Err(());
+fn verify_transcript<E: PairingEngine>(
+    cs_hash: [u8; 64],
+    contributions: &[PublicKey<E>],
+) -> Result<Vec<[u8; 64]>> {
+    let mut result = vec![];
+    let mut old_delta = E::G1Affine::prime_subgroup_generator();
+    for (i, pubkey) in contributions.iter().enumerate() {
+        let hash = hash_cs_pubkeys(cs_hash, &contributions[0..i], pubkey.s, pubkey.s_delta);
+        ensure_unchanged(
+            &pubkey.transcript[..],
+            &hash.as_ref()[..],
+            InvariantKind::Transcript,
+        )?;
+
+        // generate the G2 point from the hash
+        let r = hash_to_g2::<E>(hash.as_ref()).into_affine();
+
+        // Check the signature of knowledge
+        check_same_ratio::<E>(
+            &(pubkey.s, pubkey.s_delta),
+            &(r, pubkey.r_delta),
+            "Incorrect signature of knowledge",
+        )?;
+
+        // Check the change with the previous G1 Delta is consistent
+        check_same_ratio::<E>(
+            &(old_delta, pubkey.delta_after),
+            &(r, pubkey.r_delta),
+            "Incosistent G1 Delta",
+        )?;
+        old_delta = pubkey.delta_after;
+
+        result.push(pubkey.hash());
     }
 
-    // cs_hash should be the same
-    if &before.cs_hash[..] != &after.cs_hash[..] {
-        return Err(());
-    }
+    Ok(result)
+}
 
+#[allow(unused)]
+fn hash_params<E: PairingEngine>(params: &Parameters<E>) -> Result<[u8; 64]> {
     let sink = io::sink();
     let mut sink = HashWriter::new(sink);
-    sink.write_all(&before.cs_hash[..]).unwrap();
+    // TODO: Re-enable this
```

Comment:
This should be re-enabled once the Zexe's serialization supports writing the params.